### PR TITLE
Fix/Rosetta Rpc predecessor_id on each state change

### PR DESCRIPTION
**Pull request by firatNEAR**
Based on tests from Coinbase some of the receipts did not have `predecessor_id` in their OperationMetadata. 

This PR fixes some receipts not having `predecessor_id`

In `Rosetta RPC` for looking at block details, we only track if a `StateChange` causes change in the balance of a `StateChangeKindView::AccountTouched` in that block.

The `receipts_in_block` HashMap only includes the receipts that are included in the block. However a `StateChange` in that block could be caused by a receipt in the previous blocks. This makes `receipts_in_blocks` to not have some of the receipts that we are looking to get the `predecessor_id` for.

This change is to look for those specific receipts'  `predecessor_ids`.

### CODE REVIEW

**Review by matklad — COMMENTED** [review]
Could you elaborate a bit in the PR description why this is the right fix? 

Looking at the code, it seems that we generally have `receipts_in_block` map with recipts, and we add extra code to handle the case when the receipt isn't in that map. But that raises a question of *why* that happens? Perhaps we should fix something else to make sure all receipts are actually present in that map?

This is not a suggestion, I just genuinely don't understand why we are having this problem, and why this would be a complete and correct fix :)

**Comment by matklad** [comment]
Oh, and can we add a test which actually demonstrates the change in the behavior here, so this doesn't regress in the future?

**Review by frol — CHANGES_REQUESTED** [review]
This PR won't solve the issue, see https://github.com/near/nearcore/issues/7654#issuecomment-1253990899

**Comment by frol** [comment]
@firatNEAR  I’m not sure if this actually needs to be pushed forward. I feel having this incomplete solution where “predecessor id is not always available” is not helping much and only postpones potential problems for the users. Either you will need to push it till the very end like Indexer Framework does and ensure that predecessor id is not Option (Indexer Framework goes back to the parent blocks in the attempt to find the local delayed receipts, and if the parent is not found, it crashes), or remove this support completely in favor of #7704 (if predecessor id was only used to identify whether it was a fee transfer)

**Review by frol — APPROVED** [review]
After an offline discussion with @firatNEAR I realized that this PR will actually fix the case of the delayed receipts.

Two more corner cases will remain:

* local receipts
* delayed local receipts

Given that local receipts are only generated on transaction-to-receipt conversion (when sender-is-receiver, NEAR generates a local receipt that does not get recorded in RocksDB, so you won't be able to query it with `GetReceipt`), we may ignore these cases as those are not fees or gas refund receipts.

**Review by matklad — APPROVED** [review]

File: chain/rosetta-rpc/src/adapters/transactions.rs [hunk 1]
```diff
@@ -141,40 +141,51 @@ fn convert_cause_to_transaction_id(
...
     };
-    predecessor_id
+    Some(crate::models::AccountIdentifier::from(predecessor_id))
+}
+
+async fn get_predecessor_id_from_receipt_hash(
+    view_client: &Addr<near_client::ViewClientActor>,
+    receipt_id: CryptoHash,
+) -> Option<AccountId> {
+    let receipt_view = view_client.send(near_client::GetReceipt { receipt_id }).await.ok()?.ok()?;
```
Comment (nikurt) [inline chain/rosetta-rpc/src/adapters/transactions.rs:187]:
```suggestion
    let receipt_view = view_client.send(near_client::GetReceipt { receipt_id }.with_span_context()).await.ok()?.ok()?;
```

**Pull request merged**